### PR TITLE
WFLY-5091 User supplied externalizers are ignored by web/ejb clustering code

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-5091

It is a common misperception that since distributed web session attributes are stored in an Infinispan cache, Infinispan externalizers can be used to optimize the marshalling of session attributes.  However, in reality, user session attributes are stored within a MarshalledValue, which itself is stored in the Infinispan cache, and the clustering integration code performs marshalling lazily via JBoss Marshalling.  Therefore, the web session clustering integration code can do nothing with these externalizers.

The primary goal of this PR is to allow applications to define their own externalizers, thus allow them to both optimize their application's performance in a clustered environment, and to use non-serializable objects in the HttpSession.  These externalizers are generic to both Infinispan and JBoss Marshalling, and are adapted to either, depending on the use case.

To accomplish this, this PR splits the existing org.wildfly.clustering.marshalling module (new since WF8) into 2: org.wildfly.clustering.marshalling.api, containing public marshalling API (currently a single interface), and org.wildfly.clustering.marshalling.jboss, containing JBoss Marshalling integration code.  The new Externalizer interface (from org.wildfly.clustering.marshalling.api) is consumed by both Infinispan's marshaller, via the AdvancedExternalizerAdapter, and via JBoss Marshalling via ExternalizerObjectTable, which uses a ServiceLoader to discover instances of Externalizer accessible from a given module.

This PR also adds new marshalling configuration versions for web, sso, and ejb that leverage the new ExternalizerObjectTable.
Other improvements/changes to marshalling integration include:
- Modify all existing externalizers to implement org.wildfly.clustering.marshalling.Externalizer instead of org.wildfly.clustering.infinispan.spi.io.SimpleExternalizer (which was an extension of Infinispan's AdvancedExternalizer)
- Rename VersionedMarshallingConfiguration to MarshallingConfigurationRepository to better describe its usage
- Consolidate assorted implementations of MarshallingConfigurationRepository (formerly VersionedMarshallingConfiguration) into a single, generic implementation, i.e. SimpleMarshallingConfigurationRepository.
- Allow marshalling configuration versions to be specified via enum
- Allow ObjectTable and ClassTable instances to be configured with a specific Externalizer<Integer> for marshalling the table index. 
- IndexExternalizer contains a set of Externalizer<Integer> implementations for marshalling unsigned integer values, including a variable-length strategy, based on logic from Infinispan - useful for minimizing the serialized size of the index when the size of the object/class table is unknown.

### CODE REVIEW

File: feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/marshalling/api/main/module.xml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.3" name="org.wildfly.clustering.marshalling.api">
```
Comment (n1hility) [inline feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/marshalling/api/main/module.xml]: Did you want to mark this one as public?

Comment (pferraro) [inline feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/marshalling/api/main/module.xml]: Yes. Did I do that correctly?

**Comment by pferraro** [comment]
Ooops. Need to increment versions of new modules.

File: feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/marshalling/api/main/module.xml (hunk 1 above)
Comment (n1hility) [inline feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/marshalling/api/main/module.xml]: Sorry I see you have that in your description. Yes you did.

**Comment by stuartwdouglas** [comment]
Retest this please

**Comment by n1hility** [comment]
retest this please

**Comment by pferraro** [comment]
*sigh* New module needs another version bump.

**Comment by pferraro** [comment]
Something strange is happening - this test passes for me every time.

**Comment by pferraro** [comment]
OK - this is ready for review now.

**Pull request merged**